cognitedata/reveal
# fix(react-components): improve asset mapping cache mechanism and refactoring to stabilize rule threshold styling and switching

#### Type of change

![Bug](https://img.shields.io/badge/Type-Bug-red) 

#### Jira ticket :blue_book:
[](https://cognitedata.atlassian.net/browse/BND3D-4199)

https://cognitedata.atlassian.net/browse/BND3D-4199
https://cognitedata.atlassian.net/browse/

## Description :pencil:
This fixes the rule threshold for large rule-based + 3d model content + asset mappings by changing the asset mapping cache system in react components and some refactoring to speed up the rule threshold applying onto the 3d models, including a loading indicator when the rule is processing, loading, and applying the styling.
Before this, the asset mapping cache was based on a list of asset IDs or node IDs that were requested at once. However, the asset IDs of the requests are not always the same, forcing the cache mechanism to request the mappings even for those cached already in a different list. 
So now the cache key is based on the asset against all its mappings rather than a summed key based on a chunk of mappings.
This way it is possible to have more freedom and flexibility to use the cache without relying on a static list.

## Test instructions :information_source:

1 - Open a 3D scene in Search UI that contains contextualized assets and rule based thresholds (recommend to test in a scene that contains a huge amount of geometries and asset mappings to see the changes like Celanese, for instance).
2 - Select an Rule Based Thresholds and switch to another rule.
3. Observe the loading indicator replacing the radio button while the rule is loading and the performance improvements compared with the previous version (not crashing after loaded, more stable).

Note: it is connected to the Fusion PR https://github.com/cognitedata/fusion/pull/7149 for working with the Rule Based loading indicator

## Checklist :ballot_box_with_check:
